libs/community/langchain_community/chat_models/outlines.py

```diff
@@ -0,0 +1,564 @@
+            self.type_constraints,
+            self.json_schema,
+            self.grammar,
+        ]
+
+        num_constraints = sum(constraint is not None for constraint in constraints)
+        if num_constraints != 1 and num_constraints != 0:
+            raise ValueError(
+                "Either none or exactly one of regex, type_constraints, "
+                "json_schema, or grammar can be provided."
+            )
+        if self.regex:
+            return generate.regex(self.client, regex_str=self.regex)
+        if self.type_constraints:
+            return generate.format(self.client, python_type=self.type_constraints)
+        if self.json_schema:
+            return generate.json(self.client, schema_object=self.json_schema)
+        if self.grammar:
+            return generate.cfg(self.client, cfg_str=self.grammar)
+        return generate.text(self.client)
+
+    @property
+    def tokenizer(self) -> Any:
+        """Access the tokenizer for the underlying model.
+
+        .encode() to tokenize text.
+        .decode() to convert tokens back to text.
+        """
+        if hasattr(self.client, "tokenizer"):
+            return self.client.tokenizer
+        raise ValueError("Tokenizer not found")
+
+    def _convert_messages_to_openai_format(
+        self, messages: list[BaseMessage]
+    ) -> list[dict]:
+        return [convert_message_to_dict(message) for message in messages]
+
+    def _convert_messages_to_prompt(self, messages: list[BaseMessage]) -> str:
+        """Convert a list of messages to a single prompt."""
+        # if hasattr(self.tokenizer, "apply_chat_template"):
+        #     return self.tokenizer.apply_chat_template(
+        #         messages, tokenize=False, add_generation_prompt=True
+        #     )
+        if self.backend == "llamacpp":  # get base_model_name from gguf repo_id
+            from huggingface_hub import ModelCard
+
+            repo_creator, gguf_repo_name, file_name = self.model.split("/")
+            model_card = ModelCard.load(f"{repo_creator}/{gguf_repo_name}")
+            if hasattr(model_card.data, "base_model"):
+                model_name = model_card.data.base_model
+            else:
+                raise ValueError(f"Base model name not found for {self.model}")
+        else:
+            model_name = self.model
+
+        from transformers import AutoTokenizer
+        # from llama_cpp.llama_tokenizer import LlamaHFTokenizer
+        # LlamaHFTokenizer.from_pretrained(
+
+        return AutoTokenizer.from_pretrained(model_name).apply_chat_template(
+            self._convert_messages_to_openai_format(messages),
+            tokenize=False,
+            add_generation_prompt=True,
+        )
+
+    def bind_tools(
+        self,
+        tools: Sequence[Dict[str, Any] | type | Callable[..., Any] | BaseTool],
+        *,
+        tool_choice: Optional[Union[Dict, bool, str]] = None,
+        **kwargs: Any,
+    ) -> Runnable[
+        PromptValue
```

Comment:
can we use `from langchain_core.language_models import LanguageModelInput` instead